graph mode qat: make fake_quantize scriptable

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39828 [wip] qat graph mode: add combine_conv_bn pass
* #39837 eager qat: refactor conv-bn to make conv with scaled weights a func
* #39840 qat graph mode: make per channel observers scriptable
* #39796 qat graph mode: initial scaffolding for prepare_qat_script
* #39795 graph mode: add quantization type enum
* **#39750 graph mode qat: make fake_quantize scriptable**
* #39739 graph mode: more docs for insert observers pass
* #39683 graph mode: add docs to pre-calibration passes

Summary:

Add a test to make the default QAT qconfig scriptable, and fix
all the errors.

Test Plan:

```
python test/test_quantization.py TestQATScript.fake_quant_scriptable
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975879](https://our.internmc.facebook.com/intern/diff/D21975879)

vkuzo committed:
graph mode qat: make fake_quantize scriptable

Summary:

Add a test to make the default QAT qconfig scriptable, and fix
all the errors.

Test Plan:

```
python test/test_quantization.py TestQATScript.fake_quant_scriptable
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -59,22 +59,30 @@ def __init__(self, observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=
         self.dtype = self.activation_post_process.dtype
         self.qscheme = self.activation_post_process.qscheme
-        self.ch_axis = self.activation_post_process.ch_axis if hasattr(self.activation_post_process, 'ch_axis') else None
+        self.ch_axis = self.activation_post_process.ch_axis \
```
this is because `fake_quantize_per_channel_affine` expects an integer axis.  Axis is enforced at runtime to be >0 in that function.

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2569,3 +2570,9 @@ def forward(self, x):
                    .check_not("aten::_choose_qparams_per_tensor") \
                    .run(m1.graph)
+
+class TestQATScript(QuantizationTestCase):
```
TestQuantizeQATScript is probably better

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2569,3 +2570,9 @@ def forward(self, x):
+
+class TestQATScript(QuantizationTestCase):
+
+    def fake_quant_scriptable(self):
```
I think this test case should be in https://github.com/pytorch/pytorch/blob/master/test/quantization/test_workflow_module.py instead

jerryzh168 replied:
you can create a similar test like https://github.com/pytorch/pytorch/blob/master/test/quantization/test_workflow_module.py#L273

vkuzo committed:
Update on "graph mode qat: make fake_quantize scriptable"


Summary:

Add a test to make the default QAT qconfig scriptable, and fix
all the errors.

Test Plan:

```
python test/test_quantization.py TestQATScript.fake_quant_scriptable
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "graph mode qat: make fake_quantize scriptable"


Summary:

Add a test to make the default QAT qconfig scriptable, and fix
all the errors.

Test Plan:

```
python test/test_quantization.py TestQATScript.fake_quant_scriptable
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LG

vkuzo committed:
Update on "graph mode qat: make fake_quantize scriptable"


Summary:

Add a test to make the default QAT qconfig scriptable, and fix
all the errors.

Test Plan:

```
python test/test_quantization.py TestQATScript.fake_quant_scriptable
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975879](https://our.internmc.facebook.com/intern/diff/D21975879)

[ghstack-poisoned]

vkuzo committed:
Update on "graph mode qat: make fake_quantize scriptable"


Summary:

Add a test to make the default QAT qconfig scriptable, and fix
all the errors.

Test Plan:

```
python test/test_quantization.py TestQATScript.fake_quant_scriptable
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975879](https://our.internmc.facebook.com/intern/diff/D21975879)

[ghstack-poisoned]
